File: swift/src/Ice/PropertiesI.swift
// Copyright (c) ZeroC, Inc.

import IceImpl

final class PropertiesI: LocalObject<ICEProperties>, Properties {
    public func getProperty(_ key: String) -> String {
        handle.getProperty(key)
    }

    public func getIceProperty(_ key: String) -> String {
        handle.getIceProperty(key)
    }

    public func getPropertyWithDefault(key: String, value: String) -> String {
        handle.getPropertyWithDefault(key, value: value)
    }

    public func getPropertyAsInt(_ key: String) throws -> Int32 {
        return try autoreleasepool {
            var value: Int32 = 0
            try handle.getPropertyAsInt(key, value: &value)
            return value
        }
    }

    public func getIcePropertyAsInt(_ key: String) throws -> Int32 {
        return try autoreleasepool {
            var value: Int32 = 0
            try handle.getIcePropertyAsInt(key: key, value: &value)
            return value
        }
    }

    public func getPropertyAsIntWithDefault(key: String, value defaultValue: Int32) throws -> Int32 {
        return try autoreleasepool {
            var value: Int32 = 0
            try handle.getPropertyAsIntWithDefault(key: key, defaultValue: defaultValue, value: &value)
            return value
        }
    }

    public func getPropertyAsList(_ key: String) -> StringSeq {
        handle.getPropertyAsList(key)
    }

    public func getIcePropertyAsList(_ key: String) -> StringSeq {
        handle.getIcePropertyAsList(key)
    }

    public func getPropertyAsListWithDefault(key: String, value: StringSeq) -> StringSeq {
        handle.getPropertyAsListWithDefault(key: key, value: value)
    }

    public func getPropertiesForPrefix(_ prefix: String) -> PropertyDict {
        handle.getPropertiesForPrefix(prefix)
    }

    public func setProperty(key: String, value: String) {
        precondition(!key.isEmpty, "Key cannot be empty")
        do {
            try autoreleasepool {
                try handle.setProperty(key, value: value)
            }
        } catch {
            fatalError("\(error)")
        }
    }

    public func getCommandLineOptions() -> StringSeq {
        handle.getCommandLineOptions()
    }

    public func parseCommandLineOptions(prefix: String, options: StringSeq) throws -> StringSeq {
        try autoreleasepool {
            try handle.parseCommandLineOptions(prefix, options: options)
        }
    }

    public func parseIceCommandLineOptions(_ options: StringSeq) throws -> StringSeq {
        try autoreleasepool {
            try handle.parseIceCommandLineOptions(options)
        }
    }

    public func load(_ file: String) throws {
        try autoreleasepool {
            try handle.load(file)
        }
    }

    public func clone() -> Properties {
        PropertiesI(handle: handle.clone())
    }
}

File: python/python/Ice/PropertiesI.py
# -*- coding: utf-8 -*-
#
# Copyright (c) ZeroC, Inc. All rights reserved.
#

from .Properties import Properties

class PropertiesI(Properties):
    def __init__(self, impl):
        self._impl = impl

    def getProperty(self, key):
        return self._impl.getProperty(key)

    def getIceProperty(self, key):
        return self._impl.getIceProperty(key)

    def getPropertyWithDefault(self, key, value):
        return self._impl.getPropertyWithDefault(key, value)

    def getPropertyAsInt(self, key):
        return self._impl.getPropertyAsInt(key)

    def getIcePropertyAsInt(self, key):
        return self._impl.getIcePropertyAsInt(key)

    def getPropertyAsIntWithDefault(self, key, value):
        return self._impl.getPropertyAsIntWithDefault(key, value)

    def getPropertyAsList(self, key):
        return self._impl.getPropertyAsList(key)

    def getIcePropertyAsList(self, key):
        return self._impl.getIcePropertyAsList(key)

    def getPropertyAsListWithDefault(self, key, value):
        return self._impl.getPropertyAsListWithDefault(key, value)

    def getPropertiesForPrefix(self, prefix):
        return self._impl.getPropertiesForPrefix(prefix)

    def setProperty(self, key, value):
        self._impl.setProperty(key, value)

    def getCommandLineOptions(self):
        return self._impl.getCommandLineOptions()

    def parseCommandLineOptions(self, prefix, options):
        return self._impl.parseCommandLineOptions(prefix, options)

    def parseIceCommandLineOptions(self, options):
        return self._impl.parseIceCommandLineOptions(options)

    def load(self, file):
        self._impl.load(file)

    def clone(self):
        properties = self._impl.clone()
        return PropertiesI(properties)

    def __iter__(self):
        dict = self._impl.getPropertiesForPrefix("")
        return iter(dict)

    def __str__(self):
        return str(self._impl)
